### ISSUE TITLE
The SPRegion should infer its "inputWidth", not take it as a parameter

### ISSUE THREAD

**Issue by mrcslws** [numenta/nupic#3040]
Now that the SPRegion gets linked to multiple encoders, it's a bit clunky that we have to calculate its `inputWidth` before creating it. It should figure out the inputWidth during `initialize`.

You'd think this would be easy, but there's a problem. This `inputWidth` is also used to calculate the output size of: "spatialTopDownOut", "temporalTopDownOut", and "topDownOut". The method `getOutputElementCount` gets called **before** `initialize`. So there's no way to infer the inputWidth at that point.

This isn't just a limitation in `PyRegion`. I've tried it with C++ using a toy IdentityRegion. Output sizes can't be based on input sizes. The network API doesn't support that, you get an error that the input is not initialized. In `network::initialize` if you swap the "initialize inputs" part with the "initialize outputs" part in https://github.com/numenta/nupic.core/blob/6d2fb817b98e12eebe11d11626b8685ae8ce2d3b/src/nupic/engine/Network.cpp#L543 , it seems to solve it, but it causes other problems (lots of unit tests hit errors).

Basically I think what's needed is a iterative "Crawl the inputs and outputs, crawl the inputs and outputs, crawl the inputs and outputs, ..." repeating until everything has resolved or until we hit a point where we're not improving. We do this with network links, but not with output sizes, it seems.

**Comment by cogmission** [comment]
Marcus,

I do this (infer the input width) in HTM.java's Network API, it might be
helpful to take a look at this? Also we could discuss further if it would
be helpful?

Cheers,
David

On Mon, Mar 7, 2016 at 7:47 PM, Marcus Lewis <notifications@github.com>
wrote:

> Now that the SPRegion gets linked to multiple encoders, it's a bit clunky
> that we have to calculate its inputWidth before creating it. It should
> figure out the inputWidth during initialize.
>
> You'd think this would be easy, but there's a problem. This inputWidth is
> also used to calculate the output size of: "spatialTopDownOut",
> "temporalTopDownOut", and "topDownOut". The method getOutputElementCount
> gets called *before* initialize. So there's no way to infer the
> inputWidth at that point.
>
> This isn't just a limitation in PyRegion. I've tried it with C++ using a
> toy IdentityRegion. Output sizes can't be based on input sizes. The network
> API doesn't support that, you get an error that the input is not
> initialized. In network::initialize if you swap the "initialize inputs"
> part with the "initialize outputs" part in
> https://github.com/numenta/nupic.core/blob/6d2fb817b98e12eebe11d11626b8685ae8ce2d3b/src/nupic/engine/Network.cpp#L543
> , it seems to solve it, but it causes other problems (lots of unit tests
> hit errors).
>
> Basically I think what's needed is a iterative "Crawl the inputs and
> outputs, crawl the inputs and outputs, crawl the inputs and outputs, ..."
> repeating until everything has resolved or until we hit a point where we're
> not improving. We do this with network links, but not with output sizes, it
> seems.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/nupic/issues/3040>.
>



-- 
*With kind regards,*

David Ray
Java Solutions Architect

*Cortical.io <http://cortical.io/>*
Sponsor of:  HTM.java <https://github.com/numenta/htm.java>

d.ray@cortical.io
http://cortical.io
